File: Kattis/conquest.cpp
// https://open.kattis.com/problems/conquest

#include <bits/stdc++.h>

using namespace std;

namespace IO {
    template <typename T>
    inline void read(T &t) {
        int n = 0;
        int c = getchar_unlocked();
        t = 0;
        while (!isdigit(c)) n |= c == '-', c = getchar_unlocked();
        while (isdigit(c)) t = t * 10 + c - 48, c = getchar_unlocked();
        if (n)
            t = -t;
    }

    template <typename T, typename... Args>
    inline void read(T &t, Args &... args) {
        read(t);
        read(args...);
    }

    template <typename T>
    inline void write(T x) {
        if (x < 0)
            x = -x, putchar_unlocked('-');
        if (x > 9)
            write(x / 10);
        putchar_unlocked(x % 10 + 48);
    }

    template <typename T>
    inline void writeln(T x) {
        write(x);
        putchar_unlocked('\n');
    }
}

constexpr int MAXN = 200005;
vector<int> AL[MAXN];
int A[MAXN];
int B[MAXN];
int n,m;

int main() {

    IO::read(n,m);
    int u,v;
    while (m--) {
        IO::read(u, v);
        u--; v--;
        AL[u].emplace_back(v);
        AL[v].emplace_back(u);
    }
    for (int i = 0; i < n; ++i) IO::read(B[i]);
    int ans = B[0];
    A[0] = 1;
    set<pair<int,int>> s;
    for (const auto &i : AL[0]) {
        s.insert({B[i], i});
    }
    pair<int,int> temp;
    while (!s.empty()) {
        temp = *s.begin();
        s.erase(s.begin());
        A[temp.second] = 1;
        if (temp.first >= ans) break;
        ans += temp.first;
        for (const auto &i : AL[temp.second]) {
            if (A[i] == 0) s.insert({B[i], i});
        }
    }
    IO::writeln(ans);
    return 0;
}

File: Kattis/conquest.go
// https://open.kattis.com/problems/conquest

package main

import (
	"bufio"
	"container/heap"
	"fmt"
	"os"
)

// Fast IO Wrapper
var (
	reader *bufio.Reader = bufio.NewReader(os.Stdin)
	writer *bufio.Writer = bufio.NewWriter(os.Stdout)
)

func myPrintf(f string, a ...interface{}) { fmt.Fprintf(writer, f, a...) }
func myScanf(f string, a ...interface{})  { fmt.Fscanf(reader, f, a...) }

type Pair struct {
	first, second int
}
type EmptyStruct struct {

}

type PriorityQueue []Pair

func (pq *PriorityQueue) Len() int {
	return len(*pq)
}
func (pq *PriorityQueue) Less(i, j int) bool {
	if (*pq)[i].first == (*pq)[j].first {
		return (*pq)[i].second < (*pq)[j].second
	}
	return (*pq)[i].first < (*pq)[j].first
}
func (pq *PriorityQueue) Swap(i, j int) {
	(*pq)[i], (*pq)[j] = (*pq)[j], (*pq)[i]
}
func (pq *PriorityQueue) Push(x interface{}) {
	*pq = append(*pq, x.(Pair))
}
func (pq *PriorityQueue) Pop() interface{} {
	var n = len(*pq)
	var x = (*pq)[n-1]
	*pq = (*pq)[:n-1]
	return x
}
func (pq *PriorityQueue) Empty() bool {
	return (*pq).Len() == 0
}

var (
	n, m    int
	adjlist [][]int
	v, army []int
	freq map[Pair]EmptyStruct
	emptyStruct EmptyStruct
)

func main() {
	defer writer.Flush()
	myScanf("%d %d\n", &n, &m)
	freq = make(map[Pair]EmptyStruct)
	adjlist = make([][]int, n)
	v = make([]int, n)
	army = make([]int, n)
	var x, y int
	for i := 0; i < m; i++ {
		myScanf("%d %d\n", &x, &y)
		x--
		y--
		adjlist[x] = append(adjlist[x], y)
		adjlist[y] = append(adjlist[y], x)
	}
	for i := 0; i < n; i++ {
		myScanf("%d\n", &army[i])
	}
	var ans = army[0]
	v[0] = 1
	var pq PriorityQueue
	heap.Init(&pq)
	var pair Pair
	for _, i := range adjlist[0] {
		pair.first = army[i]
		pair.second = i
		_, ok := freq[pair]
		if !ok {
			heap.Push(&pq, pair)
			freq[pair] = emptyStruct
		}
	}
	for !pq.Empty() {
		var p = heap.Pop(&pq).(Pair)
		v[p.second] = 1
		if p.first >= ans {
			break
		}
		delete(freq, p)
		ans += p.first
		for _, i := range adjlist[p.second] {
			if v[i] == 0 {
				pair.first = army[i]
				pair.second = i
				_, ok := freq[pair]
				if !ok {
					heap.Push(&pq, pair)
					freq[pair] = emptyStruct
				}
			}
		}
	}
	myPrintf("%v\n", ans)
}
